enterprise/server/remote_execution/dirtools/dirtools.go

```diff
@@ -219,92 +216,85 @@ func newDirToUpload(instanceName string, digestFunction repb.DigestFunction_Valu
 	}
-	return &fileToUpload{
-		fullFilePath: fullFilePath,
-		info:         info,
-		resourceName: ad,
-	}, nil
 }
 
-func (f *fileToUpload) ReadSeekCloser() (io.ReadSeekCloser, error) {
-	if f.data != nil {
-		return cachetools.NewBytesReadSeekCloser(f.data), nil
-	}
-	file, err := os.Open(f.fullFilePath)
-	// Note: Caller is responsible for closing.
+// fileToUpload represents a regular file to be uploaded to cache.
+type fileToUpload struct {
+	fullPath string
+	digest   *repb.Digest
+	info     os.FileInfo
+}
+
+func newFileToUpload(digestFunction repb.DigestFunction_Value, fullPath string, info os.FileInfo) (*fileToUpload, error) {
+	d, err := digest.ComputeForFile(fullPath, digestFunction)
 	if err != nil {
 		return nil, err
 	}
-	return file, nil
+	return &fileToUpload{
+		digest:   d,
+		fullPath: fullPath,
+		info:     info,
+	}, nil
 }
 
 func (f *fileToUpload) OutputFile(rootDir string) *repb.OutputFile {
 	return &repb.OutputFile{
-		Path:         filepath.ToSlash(trimPathPrefix(f.fullFilePath, rootDir)),
-		Digest:       f.resourceName.GetDigest(),
+		Path:         filepath.ToSlash(trimPathPrefix(f.fullPath, rootDir)),
+		Digest:       f.digest,
 		IsExecutable: f.info.Mode()&0111 != 0,
 	}
 }
 
 func (f *fileToUpload) FileNode() *repb.FileNode {
 	return &repb.FileNode{
 		Name:         f.info.Name(),
-		Digest:       f.resourceName.GetDigest(),
+		Digest:       f.digest,
 		IsExecutable: f.info.Mode()&0111 != 0,
 	}
 }
 
-func (f *fileToUpload) OutputDirectory(rootDir string) *repb.OutputDirectory {
-	return &repb.OutputDirectory{
-		Path:       trimPathPrefix(f.fullFilePath, rootDir),
-		TreeDigest: f.resourceName.GetDigest(),
-	}
-}
-
-func (f *fileToUpload) DirNode() *repb.DirectoryNode {
-	return &repb.DirectoryNode{
-		Name:   f.info.Name(),
-		Digest: f.resourceName.GetDigest(),
-	}
-}
-
 func uploadFiles(ctx context.Context, uploader *cachetools.BatchCASUploader, fc interfaces.FileCache, filesToUpload []*fileToUpload) error {
 	for _, uploadableFile := range filesToUpload {
 		// Add output files to the filecache.
-		if fc != nil && uploadableFile.dir == nil {
+		if fc != nil {
 			node := uploadableFile.FileNode()
-			if err := fc.AddFile(ctx, node, uploadableFile.fullFilePath); err != nil {
+			if err := fc.AddFile(ctx, node, uploadableFile.fullPath); err != nil {
 				log.Warningf("Error adding file to filecache: %s", err)
 			}
 		}
-
-		rsc, err := uploadableFile.ReadSeekCloser()
+		f, err := os.Open(uploadableFile.fullPath)
 		if err != nil {
-			return err
+			return status.UnavailableErrorf("open output file: %s", err)
 		}
 		// Note: uploader.Upload closes the file after it is uploaded.
-		if err := uploader.Upload(uploadableFile.resourceName.GetDigest(), rsc); err != nil {
+		if err := uploader.Upload(uploadableFile.digest, f); err != nil {
 			return err
 		}
 	}
+	return nil
+}
 
+func uploadDirs(ctx context.Context, uploader *cachetools.BatchCASUploader, dirsToUpload []*dirToUpload) error {
```

Comment:
Nit: I think moving this inside UploadTree would make the whole thing easier to follow vs using a separate func.

Reply:
done